session/session.go

```diff
@@ -541,7 +542,89 @@ func (s *session) doCommit(ctx context.Context) error {
 		s.txn.SetOption(kv.KVFilter, temporaryTableKVFilter(tables))
 	}
 
-	return s.txn.Commit(tikvutil.SetSessionID(ctx, sessVars.ConnectionID))
+	return s.commitTxnWithTemporaryData(tikvutil.SetSessionID(ctx, sessVars.ConnectionID), &s.txn)
+}
+
+func (s *session) commitTxnWithTemporaryData(ctx context.Context, txn kv.Transaction) error {
+	txnTempTables := s.sessionVars.TxnCtx.TemporaryTables
+	if len(txnTempTables) == 0 {
+		return txn.Commit(ctx)
+	}
+
+	sessionData := s.sessionVars.TemporaryTableData
+	var stage kv.StagingHandle
+	for tblID, tbl := range txnTempTables {
+		if !tbl.GetModified() {
+			continue
+		}
+
+		if tbl.GetMeta().TempTableType != model.TempTableLocal {
+			continue
+		}
+
+		if sessionData == nil {
+			// Create this txn just for getting a MemBuffer. It's a little tricky
+			bufferTxn, err := s.store.BeginWithOption(tikv.DefaultStartTSOption().SetStartTS(0))
+			if err != nil {
+				return err
+			}
+
+			sessionData = bufferTxn.GetMemBuffer()
```

Comment:
It's a little tricky here, because function `newMemBuffer` and `newMemDB` are invisible, so I have to creating a new txn instance and take out its MemBuffer.

Another approach is to mody tikv/go-client to export NewMemDB as public and modify newMemBuffer to NewMemBuffer. Which one is betweer?

Reply:
You can see how it's done here https://github.com/pingcap/tidb/pull/22440/files#diff-db24cdb418822855786d681d87e4d474988173aee00b6f8d93b3a550dd4a07e3R217-R221

When the local temporary table is created, a MemDB is also created.
When the transaction commits, the data are flushed to that MemDB

https://github.com/pingcap/tidb/pull/22440/files#diff-10dab83c04a99358d431906d37f48b92be97d50a2719d4ddf81c62d3b03a0aebR538

And read need to combine the MemDB (local temporary table data inside it)  with the transaction membuffer:

https://github.com/pingcap/tidb/pull/22440/files#diff-f7f3339231e4104abed7ddd0b9a3e12d2284761165df20d2141406104051767cR329-R338

Reply:
I've studied this pr you listed, but it is a little old. KVStore has been moved to a single repo tikv/client-go.

But for the current pr, I just need a way to create a new MemDB, its not hard but should make a choice between:

1. Do not modify tikv/client-go , just use current code.
2. Modify codes in tikv/client-go, exporting function NewMemDB as public , it is still very simple.